Python 3 support, TravisCI and Coveralls integration

espenhgn opened the pull request:
Hi @mdjurfeldt, 
It would be great if you CSA implementation could be supported also with Python 3, as I now use CSA in LFPy (https://github.com/espenhgn/LFPy/tree/csa) in order to create connectivity matrices for networks set up using LFPy itself (which should support Python 2.7/3.4-3.6). 

Just a quick summary of what I did with your CSA tool:
- translated all python files using the 2to3 script, fixed some functions that would've been deprecated (`test` module, `cmp` function etc.)
- added automated testing with TravisCI integration (https://travis-ci.org/espenhgn/csa/builds) 
- Coveralls testing (when TravisCI run is successful, https://coveralls.io/github/espenhgn/csa)
- README -> README.md w. reStructuredText like formatting (to include Coveralls/TravisCI badges). 

Issues:
- the test `test_gaussnet` fails with Python 2.7 and I don't quite understand why (how come the `intervalset.__len__()` function is called at all?):
```======================================================================
ERROR: test_gaussnet (test_csa.TestElementary)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ehagen/Repositories/csa/tests/test_csa.py", line 122, in test_gaussnet
    self.assertTrue (cross (N, 0) * c)
  File "/Users/ehagen/Repositories/csa/csa/connset.py", line 252, in __mul__
    return other.__mul__ (self)
  File "/Users/ehagen/Repositories/csa/csa/connset.py", line 171, in __mul__
    return ConnectionSet (self.c.intersection (coerceCSet (other)))
  File "/Users/ehagen/Repositories/csa/csa/connset.py", line 894, in intersection
    return CSetIntersection (self, other)
  File "/Users/ehagen/Repositories/csa/csa/connset.py", line 827, in __init__
    self._mask = op1.mask ().intersection (op2)
  File "/Users/ehagen/Repositories/csa/csa/connset.py", line 289, in intersection
    return other.intersection (self)
  File "/Users/ehagen/Repositories/csa/csa/connset.py", line 568, in intersection
    return ISetBoundedMask (self.set0, self.set1, other)
  File "/Users/ehagen/Repositories/csa/csa/connset.py", line 676, in __init__
    if self.set0 and self.set1:
  File "/Users/ehagen/Repositories/csa/csa/intervalset.py", line 353, in __len__
    raise RuntimeError('ComplementaryIntervalSet has infinite length')
RuntimeError: ComplementaryIntervalSet has infinite length

----------------------------------------------------------------------
```
The tests run fine with Python 3.4-3.6, but I did not investigate further the different output. May have missed something.

mdjurfeldt commented:
Very nice! Thank you!

I think we should find the reason for the 2.7 test_gaussnet failure before merging. I can help out with this within a few days but can't look at this immediately.

espenhgn commented:
That would be awesome. If I have time, I'll try and investigate in parallel, I agree it should be fixed.

espenhgn commented:
Hi again @mdjurfeldt ,
I "fixed" the Python 2 build issue and testing on both Python 2 and 3 will now run (https://travis-ci.org/espenhgn/csa). I have no idea though of what the code change in https://github.com/INCF/csa/pull/2/commits/83e02910980f0cfd1c80b53c9a74d5d1a1d0aaf3 do though for better or worse, and would check this carefully. I also noticed that tests running on Python 3 is brittle, I do get every once in a while the test error: 
![image](https://user-images.githubusercontent.com/2492641/27945439-ff57dca8-6327-11e7-80f9-e4a79694c5ac.png)

espenhgn commented:
Hi again @mdjurfeldt .
I looked now at the code again, and decided to remove the monkey-patched `ComplementaryIntervalSet.__len__` method (commit 8191a0a), as it is not clear to me why an Exception should be raised because of infinite length of the class instantiation. Remember the class is inheriting class `IntervalSet` which has an integer length. 

Commit 4f52f19 fixes the brittle test `test_partitionRandomN`.

All tests w. TravisCI run ok: https://travis-ci.org/espenhgn/csa/

mdjurfeldt commented on csa/connset.py:
```diff
@@ -473,8 +473,7 @@ def iterator (self, low0, high0, low1, high1, state):
 def cmpPostOrder (c0, op1):
-    return cmp ((c0[1], c0[0]), (op1[1], op1[0]))
-
+    return  ((c0[1], c0[0]) > (op1[1], op1[0])) -  ((c0[1], c0[0]) < (op1[1], op1[0]))
```
Do you know if this is the most efficient implementation? I'm thinking about perhaps using some conditional instead. I don't know how that compares in Python.

espenhgn replied:
Since `cmp` is removed in Python 3.*, I found the code above as a suggested replacement. I don't know for sure, but I think it is efficient enough for the purpose. Also, the function is now removed, as `self.connections.sort(reverse=True)` appeared to give identical results as the call `self.connections.sort (cmpPostOrder)`.

mdjurfeldt replied:
Hmm... it is actually not removed since you are using it on line 503, which is correct. Using sort(reverse=True) is not correct since that gives a different ordering. (This comment does not require you to change anything---just to protest if there is some misunderstanding!)

mdjurfeldt commented on csa/_elementary.py:
```diff
@@ -125,8 +125,8 @@ def startIteration (self, state):
             obj.isPartitioned = True
-            partitions = map (self.mask.intersection, state['partitions'])
-            sizes = map (len, partitions)
+            partitions = list(map (self.mask.intersection, state['partitions']))
```
Please keep the original formatting style with space before argument list: list (...

mdjurfeldt commented on csa/_elementary.py:
```diff
@@ -234,7 +234,7 @@ def startIteration (self, state):
         if 'partitions' in state:
             obj.isPartitioned = True
-            partitions = map (self.mask.intersection, state['partitions'])
+            partitions = list(map (self.mask.intersection, state['partitions']))
```
Formatting

mdjurfeldt commented on csa/_misc.py:
```diff
@@ -181,29 +181,29 @@ def iterator (self, low0, high0, low1, high1, state):
         try:
             pre = []
-            (i, j) = maskIter.next ()
+            (i, j) = next(maskIter)
```
Formatting

mdjurfeldt commented on csa/_misc.py:
```diff
@@ -181,29 +181,29 @@ def iterator (self, low0, high0, low1, high1, state):
                 while j == post:
                     pre.append (i)
-                    (i, j) = maskIter.next ()
+                    (i, j) = next(maskIter)
```
Formatting

mdjurfeldt replied:
I'm now stopping commenting about space before argument list---please fix it everuwhere.

mdjurfeldt commented on csa/connset.py:
```diff
@@ -455,32 +455,31 @@ def __init__ (self, op1, op2):
         FiniteMask.__init__ (self)
         self.connections = list (connections)
-        self.connections.sort (cmpPostOrder)
+        self.connections.sort (reverse=True)
```
Maybe I'm missing something here, but this won't give us a post order sorting, right? Post order is _increasing_ but with the second element of a pair being most significant (as opposed to standard).

mdjurfeldt commented on csa/connset.py:
```diff
@@ -918,8 +916,7 @@ def iterator (self, low0, high0, low1, high1, state):
             ls.append ((c[1], c[0]))
-        ls.sort (cmpPostOrder)
-        return iter (ls)
+        return iter (sorted(ls))
```
I have two problems with this: 1. It doesn't give us post order. 2. It unnecessarily allocates a new list.

espenhgn replied:
fixed

espenhgn commented:
Hi. I've added now spaces (`list(` to `list (`) and similar in most places that I found, and sorted out the usage of this function `csa.connset.cmpPostOrder`. For Python 3, the function `cmp` is removed altogether and `cmp` is no longer a keyword argument to `sorted` and `list.sort`. See https://docs.python.org/3/howto/sorting.html#the-old-way-using-the-cmp-parameter for details. The rewrite of `cmpPostOrder` was based off some officially suggested replacement of the operation provided by the old `cmp` function. 

I added back the monkey-patched `csa.ComplementaryIntervalSet.__len__` method behind comments, but having it in there will still break some tests on Python 2 but not Python 3.

As for code style, I will soon provide a separate PR on that, turns out this autopep8 tool will format the code accd. to common Python practice without changing the functionality of CSA. But here, it only occluded the changes needed for Python 3 support.

mdjurfeldt commented on AUTHORS:
```diff
@@ -3,3 +3,4 @@ Maintainers of GNU Software" (maintain.texi), the section called
 "Recording Changes".
 
 Mikael Djurfeldt <mikael@djurfeldt.com>
+Espen Hagen <espenhgn at gmail.com>
```
Please add a newline. (I can't do that myself, since I can't write to your python3 branch and I don't want to mess up history by creating an extra branch.)

espenhgn replied:
Newline added! Never mind my outdated comment on "reverse=True". It should not be used anywhere now. 

Great if this could now be squashed and merged. Also, do you plan to make a new upload on PyPi?

espenhgn commented:
Excellent. Thank you @mdjurfeldt .
